subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessorScanner.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.io.CharStreams;
+import com.google.common.io.Files;
+import com.google.common.io.LineProcessor;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import org.gradle.cache.internal.FileContentCacheFactory;
+import org.gradle.internal.FileUtils;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.file.FileType;
+import org.gradle.util.DeprecationLogger;
+
+/**
+ * Discovers relevant properties of annotation processors.
+ */
+class AnnotationProcessorScanner implements FileContentCacheFactory.Calculator<Map<String, String>> {
```

Comment:
❌ Missing unit test coverage, now that this class has much more complex logic.

Reply:
Fixed.